Zend/zend_generators.c

```diff
@@ -465,10 +540,13 @@ static void zend_generator_add_single_child(zend_generator_node *node, zend_gene
 			node->child.ht = ht;
 		}
 
-		zend_hash_index_add_ptr(node->child.ht, (zend_ulong) leaf, child);
+		if (zend_hash_index_add_ptr(node->child.ht, (zend_ulong) leaf, child) == NULL) {
+			ZEND_ASSERT(node->children > 1);
```

Comment:
Hm, I don't quite get under what circumstances the leaf is already in the table?

Reply:
It may be added a second time while merging child nodes, when the parent is already a multi_children_node. Could be probably also prevented by tweaking the conditions in the final while loop of zend_generator_add_child(), but I didn't bother figuring that out with respect to all possible cases we have there.

Reply:
If you want to have a look, check out Zend/tests/generators/yield_from_multi_tree.phpt

Reply:
Ah, I remember the exact reason: each node with just one direct child does not have all the eventual leafs  … now, when we add a child to a non-leaf node we now have two children - and in that case we need to add all the children of this nodes next multi_children_node, which also contains the current leaf.
We need to skip that one at least.